### ISSUE TITLE
sublist limit

### ISSUE THREAD

**Issue by flycoolman** [netconf-wg/list-pagination#43]
I agree with this MVP approach but I do not understand the sublist-limit.
Our server supports the 'depth' parameter but not a sublist limit.

  grouping sublist-limit-param-grouping {
    description
      "This grouping may be used by protocol-specific YANG modules
       to define a protocol-specific query parameter.";
    leaf sublist-limit {
      type union {
        type uint32 {
          range "1..max";
        }
        type enumeration {
          enum "unbounded" {
            description
              "Indicates that the number of entries that may be
               returned is unbounded.";
          }
        }
      }
      default "unbounded";
      description
        "The 'sublist-limit' parameter limits the number of entries
         for descendent lists and leaf-lists.

         Any result-set that is limited includes, somewhere in
         its encoding, the metadata value 'remaining' to indicate
         the number entries not included in the result set.";
    }
  }

1) The NC and RC drafts do not provide any additional information for this
parameter or the target list or leaf-list
Maybe they do not need to say anything, but it is very unclear how the NC
XPath or subtree filter, and RC target resource
produce a single list or leaf-list for pagination.  It is unclear what
interoperable filters a client can use on all servers.

2)  This sub-list limit applies to all descendant nodes. The server is
supposed to keep a running count for all descendent lists and leaf-lists
 It seems like this filter would be hard to implement.  IMO, this parameter
should be optional.

3) leaf-lists do not have child nodes. The first description sentence means
if the target list
has descendant leaf-lists, they count towards the subentry limit.  Where
are these large leaf-lists that
need so much special processing?  IMO descendant leaf-lists should not be
counted in the limit.

**Comment by billwuqin** [comment]
From Andy Bierman:
https://mailarchive.ietf.org/arch/msg/netconf/3xUYErnG9FHRnMVzpunZTKxDtpM/

**Comment by flycoolman** [comment]
Comment 1:
I don't think it is needed to mention the details of the sublist-limit again in NC or RC.
Comment 2:
It looks like the server needs to know the number of the descendent lists and leaf-lists.
Comment 3:
Any ideas?

**Comment by avtobiff** [comment]
I wonder if we should leave this as is,
at least for the time being.

We should have engaged in discussion
with Andy earlier...

**Comment by avtobiff** [comment]
I believe this is answered on the list. [0]

Answer to question 3:

> We have a similar feature implemented in a different API for lists and
leaf-lists.
When building UIs it can be convenient to have this sublist-limit for
leaf-lists.

[0] https://mailarchive.ietf.org/arch/msg/netconf/I6ziQ9qzeCy0WcNBWnto1RNVpCE/

**Comment by billwuqin** [comment]
> 
> 1. The NC and RC drafts do not provide any additional information for this
>    parameter or the target list or leaf-list
>    Maybe they do not need to say anything, but it is very unclear how the NC
>    XPath or subtree filter, and RC target resource
>    produce a single list or leaf-list for pagination.  It is unclear what
>    interoperable filters a client can use on all servers.
I think Lada has similar comment as follows:
"
It is referred to sec. 6.4.1 of RFC 7950 for the definition of XPath context.
However, that section doesn't define the context node, so section 3.1.1 should
state that the context node is the target of the query (I assume). Moreover,
the accessible tree as defined in sec. 6.4.1 of RFC 7950 depends on
circumstances that are specific to YANG, so I think an explicit definition of
the accessible tree should be given in sec. 3.1.1.
"
In other words, context node and accessible tree needs to be explicitly specified.

**Comment by flycoolman** [comment]
> 1. In other words, context node and accessible tree needs to be explicitly specified.

@billwuqin have added this in issue #45

**Comment by kwatsen** [comment]
There should be -nc and -rc examples showing the sublist-limit parameter being used.  Let's add if missing.

There must be some misunderstanding with regards to implementation-complexity.  The sublist-limit parameter, whilst set at the query-level, is evaluated on a list by list basis (no state needs to be maintained across lists).  For instance, assuming all dependent lists have millions of elements, with sublist-limit=3, only the first three elements of every (potentially recursive list) contains three elements in the response.

**Comment by billwuqin** [comment]
> I believe this is answered on the list. [0]
> 
> Answer to question 3:
> 
> > We have a similar feature implemented in a different API for lists and
> > leaf-lists.
> > When building UIs it can be convenient to have this sublist-limit for
> > leaf-lists.
> 
> [0] https://mailarchive.ietf.org/arch/msg/netconf/I6ziQ9qzeCy0WcNBWnto1RNVpCE/

I think we only support leaf-list is nested in list but, we dont' support leaf-list is nested in leaf-list, therefore Andy's statement "Leaf-list doesn't have child node" is correct.

**Comment by billwuqin** [comment]
> There should be -nc and -rc examples showing the sublist-limit parameter being used. Let's add if missing.
> 
> There must be some misunderstanding with regards to implementation-complexity. The sublist-limit parameter, whilst set at the query-level, is evaluated on a list by list basis (no state needs to be maintained across lists). For instance, assuming all dependent lists have millions of elements, with sublist-limit=3, only the first three elements of every (potentially recursive list) contains three elements in the response.

Agree with all your comments, I see Andy's follow up comment to Question 1 as follows:
> Do you mean the "where" XPath expressions?
>

No.

I don't see how this parameter works together with the offset + limit
approach.
The cursor approach could probably work.

Consider a simple YANG subtree

   container top {
      list listA {
         key A;
         leaf A { type int32; }
         list listB {
           key B;
           leaf B { type int32; }
         }
       }
    }

Assume listA has 2 entries and each nested listB has 3 entries.
If the target is /top/listA  and sublist-limit=2 then what does the "walk"
look like
to retrieve the target list?

It seems like the cursor approach could return the last sub-list.
Not so clear how the offset approach handles a partial target list entry.

He seems to have some suspicious on query parameter combination "offset+limit", but don't have doubt on cursor+limit

**Comment by billwuqin** [comment]
The best way to move forward is to keep as it is, or we can introduce similar 'sort' feature approach to make sublist-limit optional? No?

**Comment by avtobiff** [comment]
I think we should move forward as is.

**Comment by avtobiff** [comment]
Closing this
